Started working on achievements

efeka opened the pull request:
I added a few achievements related to total score, total lines cleared and number of lines cleared at once. When the player completes an achievement, it is now displayed on the screen similar to the bonus score notification.

efeka committed:
started working on achievements

efeka committed:
achievements are now being displayed

efeka committed:
achievements are now being displayed

cagdasgerede commented on src/main/java/com/quirko/app/GameController.java:
```diff
@@ -27,6 +30,10 @@ public DownData onDownEvent(MoveEvent event) {
             board.mergeBrickToBackground();
             clearRow = board.clearRows();
+
+            //check line clear related achievements
```
gereksiz comment

cagdasgerede commented on src/main/java/com/quirko/app/GameController.java:
```diff
@@ -39,9 +46,11 @@ public DownData onDownEvent(MoveEvent event) {
         } else {
             if (event.getEventSource() == EventSource.USER) {
                 board.getScore().add(1);
+                //check score related achievements
```
gereksiz comment

cagdasgerede commented on src/main/java/com/quirko/gui/GuiController.java:
```diff
@@ -239,6 +239,13 @@ private void moveDown(MoveEvent event) {
                 groupNotification.getChildren().add(notificationPanel);
                 notificationPanel.showScore(groupNotification.getChildren());
             }
+            //display achievements when they are completed
```
gereksiz comment

cagdasgerede commented on src/main/java/com/quirko/logic/Achievements.java:
```diff
@@ -0,0 +1,72 @@
+            return completed.peek();
+    }
+
+}
```
missing newline

cagdasgerede commented on src/main/java/com/quirko/logic/DownData.java:
```diff
@@ -3,10 +3,12 @@
 public final class DownData {
     private final ClearRow clearRow;
     private final ViewData viewData;
+    private final Achievements achievement;
```
```Achievements achivements```

cagdasgerede commented on src/main/java/com/quirko/logic/Achievements.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    public void clearedLines(int lines) {
+        totalLinesCleared += lines;
```
Bu toplam rakam, uygulama her acilip kapandiginda sifirlanir. Bu achievementlari bir dosyaya persist etmen lazim. Her uygulama acildiginda da bu achievementlari okuman lazim. Aksi halde bu achievementlar sadece her yeni oyun acilisi boyunca gecerli olur.

cagdasgerede commented on src/main/java/com/quirko/logic/Achievements.java:
```diff
@@ -0,0 +1,72 @@
+import java.util.Queue;
+import java.util.Stack;
+
+public class Achievements {
```
Bu classtan subclasslar yaz. Her subclassin adi ne tur bir achievement ise onu yansitsin.

Ortak kodlar Achivement classinda olsun. Achivement classinda isAchieved gibi boolean donen bir metot tanimlanabilir. Subclasslar bunu override edebilir.

cagdasgerede reviewed (changes requested):
Achivement lari bir class hiyerarsisi icerisinde tanimla. Achievement diye bir class olsun ve specific achievementlar bu classin sublcasslari olsunlar. Achivementlar oyun kapatilip acilinca da dogru track edilmeye devam etsinler. Bunun icin achivementlari bir dosyaya persist etmen lazim.

Her achivement tipi icin unit testler yaz.

Ayrica su ana kadar yaptigin degisiklikleri anlatan bir videonun linkini burada paylas.

efeka committed:
organized achievements

efeka committed:
small adjustment on achievement milestones

efeka committed:
each users achievements can now be saved and loaded from a file

For this system to work properly, users need to be able to enter their usernames when the game is launched. I will be implementing that soon.

efeka commented:
Hocam bir sonraki pushumda unit testler de olacak. İstediğiniz videoyu henüz çekemedim, şimdi kodun söylediğiniz zamandaki versiyonuna göre mi çekeyim yoksa şimdiki haline göre mi? Yoksa ikisi birden mi?

cagdasgerede commented on src/main/java/com/quirko/logic/achievements/AchievementFileIO.java:
```diff
@@ -0,0 +1,110 @@
+
+    private void loadAchievementData() {
+        int line = 0;
+        try {
```
fix indentation

cagdasgerede commented on src/main/java/com/quirko/logic/achievements/AchievementFileIO.java:
```diff
@@ -0,0 +1,110 @@
+                line++;
+            }
+            reader.close();
+		} catch (Exception e) {
```
fix indentation

cagdasgerede commented on src/main/java/com/quirko/logic/achievements/AchievementFileIO.java:
```diff
@@ -0,0 +1,110 @@
+            writer.close();
+                
+        } catch(Exception e) {
+            e.printStackTrace();
```
Bu tür exceptionları log dosyasına logla.

cagdasgerede commented on src/main/java/com/quirko/logic/achievements/AchievementFileIO.java:
```diff
@@ -0,0 +1,110 @@
+            userAchievements.set(3 * index, achievements.totalCleared.getLastCompleted() + "");
+            userAchievements.set(3 * index + 1, achievements.clearedAtOnce.getLastCompleted() + "");
+            userAchievements.set(3 * index + 2, achievements.scored.getLastCompleted() + "");
+            System.out.println(userAchievements);
```
kodda sysout görmek istemiyorum.

cagdasgerede commented on src/main/java/com/quirko/logic/achievements/Achievements.java:
```diff
@@ -0,0 +1,13 @@
+    public boolean checkAchievement(int number);
+    public int getLastCompleted();
+    
+}
```
missing newline

cagdasgerede commented on src/main/java/com/quirko/logic/achievements/TotalLinesCleared.java:
```diff
@@ -0,0 +1,61 @@
+            return completed.peek();
+        return 0;
+    }
+}
```
missing newline

cagdasgerede commented on src/main/resources/user_achievements.txt:
```diff
@@ -0,0 +1,24 @@
+a
```
Bu dosyayı repoya eklemeye gerek var mı?

cagdasgerede reviewed (changes requested):
Genel olarak achivement hiyerarşisi uygun.

Yaptıklarını görsel olarak gösteren bir videoyu bekliyorum halen. Aynı zamanda kodun için herhangi bir test yazmamışsın. Bir sonraki commitinde manalı unit testler de ekle lütfen.

Review etmemi istediğinde bana assign etmeyi unutma.

efeka commented:
Merhaba hocam, istediğiniz videoyu çektim.
https://www.youtube.com/watch?v=y2KKuR5ztfU&ab_channel=efekaraman

efeka committed:
updated .gitignore

efeka committed:
added missing newlines

efeka committed:
fixed indentations

efeka committed:
added username input window

This window comes up when the application is started. Game starts after the user saves their username.

efeka committed:
fixed problems with file IO, completed the relation between achievements and usernames

There was a problem with updating user_achievements.txt file when adding new users, which is fixed now.

efeka committed:
fixed bug where getting a high tier achievement wouldn't automatically complete lower tiers

efeka committed:
wrote unit tests for each type of achievement

efeka committed:
wrote unit tests for each type of achievement

efeka committed:
removed sysout

efeka committed:
updated .gitignore

efeka committed:
added MyAchievementsWindow

efeka committed:
improved user name input window

efeka committed:
added a window where user can view their own achievements

efeka committed:
small gui improvement
